File: apps/shim-python/loopstorm/_connection.py
# SPDX-License-Identifier: MIT
"""UDS connection management for the LoopStorm engine."""

from __future__ import annotations

import logging
import os
import socket
import sys

from loopstorm._errors import (
    ConnectionClosedError,
    EngineUnavailableError,
    MessageTooLargeError,
)
from loopstorm._protocol import DecisionRequest, DecisionResponse

_MAX_MESSAGE_SIZE = 65_536  # 64 KiB
_RECV_CHUNK = 4096

logger = logging.getLogger("loopstorm")


def resolve_socket_path(socket_path: str | None) -> str:
    """Resolve the engine socket path from arg, env, or platform default."""
    if socket_path is not None:
        return socket_path
    env = os.environ.get("LOOPSTORM_SOCKET")
    if env:
        return env
    if sys.platform == "win32":
        return r"\\.\pipe\loopstorm-engine"
    return "/tmp/loopstorm-engine.sock"


class EngineConnection:
    """Manages a UDS connection to the loopstorm-engine process."""

    def __init__(self, socket_path: str, timeout: float) -> None:
        self._socket_path = socket_path
        self._timeout = timeout
        self._sock: socket.socket | None = None
        self._buffer = b""

    def _connect(self) -> None:
        """Establish the UDS connection."""
        if not hasattr(socket, "AF_UNIX"):
            raise EngineUnavailableError(
                "Unix domain sockets are not supported on this platform"
            )
        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.settimeout(self._timeout)
        sock.connect(self._socket_path)
        self._sock = sock
        self._buffer = b""

    def _ensure_connected(self) -> None:
        """Lazy connect: establish connection if not already connected."""
        if self._sock is None:
            self._connect()

    def send_request(self, request: DecisionRequest) -> None:
        """Serialize request as NDJSON and send over UDS."""
        self._ensure_connected()
        line = request.to_ndjson()
        if len(line) > _MAX_MESSAGE_SIZE:
            raise MessageTooLargeError(len(line))
        assert self._sock is not None  # ensured by _ensure_connected
        self._sock.sendall(line)

    def recv_response(self) -> DecisionResponse:
        """Read one NDJSON line from the socket and deserialize."""
        assert self._sock is not None
        while b"\n" not in self._buffer:
            chunk = self._sock.recv(_RECV_CHUNK)
            if not chunk:
                raise ConnectionClosedError()
            self._buffer += chunk
            if len(self._buffer) > _MAX_MESSAGE_SIZE:
                raise MessageTooLargeError(len(self._buffer))
        idx = self._buffer.index(b"\n")
        line = self._buffer[:idx]
        self._buffer = self._buffer[idx + 1 :]
        return DecisionResponse.from_json(line.decode("utf-8"))

    def request(self, req: DecisionRequest) -> DecisionResponse:
        """Send a request and return the response."""
        self.send_request(req)
        return self.recv_response()

    def reconnect(self) -> None:
        """Close and reestablish the connection."""
        self.close()
        self._connect()

    def close(self) -> None:
        """Close the UDS connection."""
        if self._sock is not None:
            try:
                self._sock.close()
            except OSError:
                pass
            self._sock = None
            self._buffer = b""

    @property
    def connected(self) -> bool:
        """Return True if a socket connection exists."""
        return self._sock is not None

    def fileno(self) -> int:
        """Return the socket file descriptor (for testing)."""
        if self._sock is None:
            raise ValueError("not connected")
        return self._sock.fileno()

File: apps/shim-ts/src/connection.ts
// SPDX-License-Identifier: MIT
/**
 * IPC connection management for the LoopStorm engine.
 *
 * Uses Node.js net.Socket for Unix Domain Socket (Linux/macOS) and
 * named pipe (Windows) connections. Both are accessed via the same
 * net.Socket API with a `path` option.
 */

import { Socket } from "node:net";
import { ConnectionClosedError, EngineUnavailableError, MessageTooLargeError } from "./errors.js";
import { type DecisionRequest, DecisionResponse } from "./protocol.js";

const MAX_MESSAGE_SIZE = 65_536; // 64 KiB

/**
 * Resolve the engine socket path from arg, env, or platform default.
 */
export function resolveSocketPath(socketPath?: string | undefined): string {
  if (socketPath) return socketPath;
  const envPath = process.env.LOOPSTORM_SOCKET;
  if (envPath) return envPath;
  if (process.platform === "win32") return "\\\\.\\pipe\\loopstorm-engine";
  return "/tmp/loopstorm-engine.sock";
}

/**
 * Manages an IPC connection to the loopstorm-engine process.
 */
export class EngineConnection {
  private readonly socketPath: string;
  private readonly timeoutMs: number;
  private socket: Socket | null = null;
  private buffer: Buffer = Buffer.alloc(0);

  constructor(socketPath: string, timeoutMs: number) {
    this.socketPath = socketPath;
    this.timeoutMs = timeoutMs;
  }

  /** Establish the IPC connection. */
  private connect(): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      const sock = new Socket();
      sock.setTimeout(this.timeoutMs);

      const onError = (err: Error) => {
        sock.removeListener("error", onError);
        sock.destroy();
        reject(new EngineUnavailableError(err.message));
      };

      sock.on("error", onError);
      sock.connect({ path: this.socketPath }, () => {
        sock.removeListener("error", onError);
        this.socket = sock;
        this.buffer = Buffer.alloc(0);
        resolve();
      });
    });
  }

  /** Lazy connect: establish connection if not already connected. */
  private async ensureConnected(): Promise<void> {
    if (this.socket === null) {
      await this.connect();
    }
  }

  /** Serialize request as NDJSON and send over IPC. */
  async sendRequest(request: DecisionRequest): Promise<void> {
    await this.ensureConnected();
    const line = request.toNdjson();
    if (line.length > MAX_MESSAGE_SIZE) {
      throw new MessageTooLargeError(line.length);
    }
    return new Promise<void>((resolve, reject) => {
      this.socket?.write(line, (err?: Error | null) => {
        if (err) reject(new EngineUnavailableError(err.message));
        else resolve();
      });
    });
  }

  /** Read one NDJSON line from the socket and deserialize. */
  recvResponse(): Promise<DecisionResponse> {
    return new Promise<DecisionResponse>((resolve, reject) => {
      const sock = this.socket;
      if (!sock) {
        reject(new EngineUnavailableError("not connected"));
        return;
      }

      // Check if we already have a complete line buffered
      const existingIdx = this.buffer.indexOf(0x0a);
      if (existingIdx >= 0) {
        const line = this.buffer.subarray(0, existingIdx).toString("utf-8");
        this.buffer = this.buffer.subarray(existingIdx + 1);
        resolve(DecisionResponse.fromJson(line));
        return;
      }

      const cleanup = () => {
        sock.removeListener("data", onData);
        sock.removeListener("error", onError);
        sock.removeListener("close", onClose);
        sock.removeListener("timeout", onTimeout);
      };

      const onData = (chunk: Buffer) => {
        this.buffer = Buffer.concat([this.buffer, chunk]);
        if (this.buffer.length > MAX_MESSAGE_SIZE) {
          cleanup();
          reject(new MessageTooLargeError(this.buffer.length));
          return;
        }
        const idx = this.buffer.indexOf(0x0a);
        if (idx >= 0) {
          const line = this.buffer.subarray(0, idx).toString("utf-8");
          this.buffer = this.buffer.subarray(idx + 1);
          cleanup();
          resolve(DecisionResponse.fromJson(line));
        }
      };

      const onError = (err: Error) => {
        cleanup();
        reject(new EngineUnavailableError(err.message));
      };

      const onClose = () => {
        cleanup();
        reject(new ConnectionClosedError());
      };

      const onTimeout = () => {
        cleanup();
        reject(new EngineUnavailableError("socket timeout"));
      };

      sock.on("data", onData);
      sock.on("error", onError);
      sock.on("close", onClose);
      sock.on("timeout", onTimeout);
    });
  }

  /** Send a request and return the response. */
  async request(req: DecisionRequest): Promise<DecisionResponse> {
    await this.sendRequest(req);
    return this.recvResponse();
  }

  /** Close and reestablish the connection. */
  async reconnect(): Promise<void> {
    this.close();
    await this.connect();
  }

  /** Close the IPC connection. */
  close(): void {
    if (this.socket !== null) {
      try {
        this.socket.destroy();
      } catch {
        // ignore
      }
      this.socket = null;
      this.buffer = Buffer.alloc(0);
    }
  }

  /** True if a socket connection exists. */
  get connected(): boolean {
    return this.socket !== null;
  }
}
